src/pkg/runner/actions.go

```diff
@@ -264,17 +272,60 @@ func GetBaseActionCfg[T any](cfg types.ActionDefaults, a types.BaseAction[T], va
 		cfg.Dir = *a.Dir
 	}
 
-	if len(a.Env) > 0 {
-		cfg.Env = append(cfg.Env, a.Env...)
-	}
-
 	if a.Shell != nil {
 		cfg.Shell = *a.Shell
 	}
 
-	// Add variables to the environment.
+	defaultEnv := cfg.Env
+
+	var combinedEnv map[string]string = make(map[string]string)
+
+	// Add the sets of variables to the combined environment in the following order of priority:
+	// 1 - anything from config.GetExtraEnv()
+	// 2 - any vars passed in. These are the --set variables
+	// 3 - any Env from the action
+	// 4 - any default environment from types.ActionDefaults
+	//
+	// Each one will only add the key if it is not already present, enforcing the above order of precedence.
+
+	for k, v := range config.GetExtraEnv() {
+		combinedEnv[k] = v
+	}
+
 	for k, v := range vars {
-		cfg.Env = append(cfg.Env, fmt.Sprintf("%s=%s", k, v.Value))
+		if _, exists := combinedEnv[k]; !exists {
+			combinedEnv[k] = v.Value
+		}
+	}
+
+	for _, envVar := range a.Env {
+		parts := strings.SplitN(envVar, "=", 2)
+		if len(parts) == 2 {
+			k := parts[0]
+			v := parts[1]
+
+			if _, exists := combinedEnv[k]; !exists {
+				combinedEnv[k] = v
+			}
+		}
+	}
+
+	for _, envVar := range defaultEnv {
+		parts := strings.SplitN(envVar, "=", 2)
+		if len(parts) == 2 {
+			k := parts[0]
+			v := parts[1]
+
+			if _, exists := combinedEnv[k]; !exists {
+				combinedEnv[k] = v
+			}
+		}
+	}
+
+	// format all the combinedEnv back into "KEY=VALUE" strings and put into cfg.Env
+	cfg.Env = nil
+	for k, v := range combinedEnv {
+		cfg.Env = append(cfg.Env, fmt.Sprintf("%s=%s", k, v))
```

Comment:
I think these could still be duped we just would have to find the order these go in to have the right precedence